Validate that shape dimension is either int or float value

abhishekgarg16 opened the pull request:
Fix for bug [#299](https://github.com/pydy/pydy/issues/299)

Dimensions passed as attributes to shapes should either be integer or float values. I have added a function check_valid_input() which checks that the dimensions passed are either integer or float values. If it is not so, it raises 'TypeError' in that case.
- [ ] There are no merge conflicts.
- [ ] If there is a related issue, a reference to that issue is in the
  commit message.
- [ ] Unit tests have been added for the new feature.
- [ ] The PR passes tests both locally (run `nosetests`) and on Travis CI.
- [ ] All public methods and classes have docstrings. (We use the [numpydoc
  format](https://github.com/numpy/numpy/blob/master/doc/HOWTO_DOCUMENT.rst.txt).)
- [ ] An explanation has been added to the online documentation. (`docs`
  directory)
- [ ] The code follows PEP8 guidelines. (use a linter, e.g.
  [pylint](http://www.pylint.org), to check your code)
- [ ] The new feature is documented in the [Release
  Notes](https://github.com/pydy/pydy#release-notes).
- [ ] The code is backwards compatible. (All public methods/classes must
  follow deprecation cycles.)
- [ ] All reviewer comments have been addressed.

moorepants commented:
You'll need to write some units tests that ensure this doesn't regress. The tests should try both correct and incorrect input.

moorepants commented:
What happens if I do this:

```python
s = Cylinder()
s.radius = 'xyz'
```

moorepants commented:
You also need to run all of the tests locally before pushing. Note that the Ci builds fail a bunch of tests with your new error.

abhishekgarg16 commented:
@moorepants , but when we create an instance of cylinder shape, we need to pass its length and radius. Why do you want to do it as 
`s = Cylinder()`
`s.radius = 'xyz'`

abhishekgarg16 commented:
The code written by me follows the PEP-8 standards.
Can you please help me figure out why all the tests have failed?

abhishekgarg16 commented:
@moorepants , the tests for the shape module were passing strings as shape dimensions. I have modified the tests to have integers or float values as shape dimensions. I have also added tests for 'TypeError' in case the dimensions passed are not integer or float values. I ran all the tests on my local machine and they all passed.

abhishekgarg16 commented:
The issue raised by you in your second comment is valid. If the attributes of an instance are changed after its creation, 'TypeError' is not raised in that case. I am looking into what can be done to fix this. Any idea, how to solve this problem?

moorepants commented:
This check should happen in the propert setter, not in the `__init__`.

abhishekgarg16 commented:
@moorepants , I have added property setters in each class to verify the same. I was having problems while trying to do it in the parent class, since the number of dimensions is different for each shape. All the tests have passed on my local machine.

abhishekgarg16 commented:
@moorepants , I have added the property setters in the parent class 'Shape'. Also, i have changed some tests in the test_shapes.py file to adapt to the new code. These tests seem to be failing here because the tests themselves are wrong according to the new fix and need to be changed.

moorepants commented on pydy/viz/tests/test_scene.py:
```diff
@@ -59,8 +59,7 @@ def setup(self):
         self.viz_frame_sym_shape = VisualizationFrame(ceiling, block,
-                                                      Sphere(radius=mass /
-                                                             10.0))
+                                                      Sphere(radius = 10.0))
```
These tests are correct. We allow symbols to be passed in as the attributes for shapes so that you can define them symbolically. Please don't change the current tests unless you have very good reason to.

moorepants commented on pydy/viz/tests/test_scene.py:
```diff
@@ -250,7 +249,7 @@ def test_generate_scene_dict(self):
         scene._generate_scene_dict()
 
-        assert scene._scene_info['objects'][viz_frame_id]['radius'] == 0.1
+        assert scene._scene_info['objects'][viz_frame_id]['radius'] == 10.0 
```
Do not change tests.

moorepants commented on pydy/viz/tests/test_shapes.py:
```diff
@@ -57,30 +57,31 @@ def test_shape():
 
     shape = Shape()
-    shape.length = symbols('l')
+    shape.length = 16.0
```
All of the changes in this file need to be reverted. These tests specifically test the symbolic functionality for the shape dimensions.

moorepants commented:
I'm sorry but all of this is not correct. The shape dimensions can be numerical values or sympy symbolics.

oliverlee commented on pydy/viz/shapes.py:
```diff
@@ -679,7 +718,10 @@ def tube_radius(self):
     @tube_radius.setter
     def tube_radius(self, new_tube_radius):
-        self._tube_radius = new_tube_radius
+        if not isinstance(new_tube_radius, int) and not isinstance(new_tube_radius, float):
```
`isinstance()` allows for a tuple of object types as the second argument.

moorepants commented:
Closing as this seems abandoned by the submitter.
